lib/Request.js

```diff
@@ -124,101 +133,124 @@ Request.prototype.setQuery = function(query, value) {
     return this;
 };
 
+
+Request.prototype.deleteQuery = function(query) {
+    delete this._url.query[query];
+    return this;
+};
+
 Request.prototype.getQuery = function(query) {
     return this._url.query.hasOwnProperty(query) ? this._url.query[query] : undefined;
 };
 
+Request.prototype.validateAndCastParams = function(object) {
+    return this.validateAndCast(this.getParams());
+}
+
+Request.prototype.validateAndCastQueries = function(object) {
+    return this.validateAndCast(this.getQueries());
+
+}
+
 /**
- * Validates the query with the provided schema.
+ * Validates the query  or passed in params with the provided schema.
  * Query parameters that are not included in the schema are ignored.
- * Query parameters that are defined as type "integer" or "number" are converted accordingly.
- * Returns a validation object with two properties:
- *   - boolean valid -- true if the query parameters are valid
- *   - array errors -- array of properties that are in error, if any
+ * Query parameters that are defined as type "integer", "number", "boolean" or "string" are converted accordingly.
+ * Pushes errors to validationErrors array if errors are gound
  *
- * @param object schema - a schema in json-schema format.
- * @return object - validation result
+ * @param object - object to validate
  */
-Request.prototype.validateQuery = function(schema, defaults) {
-    // Nothing to validate if the schema is not an object (rare)
-    if (!schema.properties) {
-        // Clear out existing query params -- they must be in the schema to access them.
-        this.setQueries({});
-        return {
-            valid: true,
-            errors: []
-        };
+Request.prototype.validateAndCast = function(object) {
+    if (!Object.keys(object).length) {
+        return;
     }
 
-    // Build new query object based on defaults
-    var query = _.assign({}, defaults);
-
-    // Validate query with schema
-    var validate = jsen(schema);
+    for (var key in object) {
+        var value = object[key];
 
-    for (var key in query) {
-        // Skip prototype properties, if any
-        if (!query.hasOwnProperty(key)) continue;
+        var filter = new SchemaFilter(this._schema);
+        var schemaProperty = filter.property(key);
 
         // Skip properties that are not in the schema
-        if (!schema.properties.hasOwnProperty(key)) continue;
-
-        if (defaults[key] === '') {
-            // Empty string in defaults means no default, i.e. undefined
-            query[key] = undefined;
+        if (!schemaProperty) {
+            throw new Error("Unexpected query param " + key + " not found in schema.");
         }
 
-        var value = this.getQuery(key);
-        if (typeof value === 'undefined' || value === '') {
+        //If null are allowed and the value is null than don't transform the value
+        if (value == null && schemaProperty.type.indexOf('null') !== -1) {
```

Comment:
`(stringOrArray).indexOf(someString) !== -1` may potentially lead to false positives if the value is a string and contains the string `"null"`, e.g. `nullster`. Since JSON schema supports custom types, I think we should be a bit more strict in our check.

    if property type is a string
      do strict equality check
    else if property is an array
      do indexOf check
    else
      what to do if type is provided?